### PR TITLE
Fix the TF Trainer gradient accumulation and the TF NER example

### LINKED ISSUE #6479: [TFTrainer] gradient accumulation error
## Environment info
     
- `transformers` version: master (#9a8c168)
- Tensorflow version: 2.3.0

### Who can help 

 Trainer: @sgugger
 tensorflow: @jplu 

## Information

When using >1 `gradient_accumulation_steps` with TFTrainer and model inputs which are *not* simple tensors (for example dicts) the trainer fails. Also, to me it looks like there are logic issues in the way the `reduced_features` are computed for the gradient accumulation (not sure though).

Issue is here: https://github.com/huggingface/transformers/blob/9a8c168f56fe3c0e21d554a577ac03beb004ef89/src/transformers/trainer_tf.py#L602

## To reproduce

```python
import tensorflow as tf
from transformers import TFT5ForConditionalGeneration, TFTrainer, TFTrainingArguments

input_ids = [[1, 2, 3], [1, 2, 3]]
labels = [1, 2, 3]

dataset = tf.data.Dataset.from_tensor_slices(({'input_ids': input_ids}, labels))
model = TFT5ForConditionalGeneration.from_pretrained("t5-base")

training_args = TFTrainingArguments(
    output_dir='./results',  # output directory
    logging_steps=100,
    max_steps=2,
    save_steps=2000,
    per_device_train_batch_size=2,  # batch size per device during training
    per_device_eval_batch_size=8,  # batch size for evaluation
    warmup_steps=0,  # number of warmup steps for learning rate scheduler
    weight_decay=0.0,  # strength of weight decay
    learning_rate=5e-5,
    gradient_accumulation_steps=2
)

with training_args.strategy.scope():
    model = TFT5ForConditionalGeneration.from_pretrained("t5-base")

trainer = TFTrainer(
    model=model,  # the instantiated 🤗 Transformers model to be trained
    args=training_args,  # training arguments, defined above
    train_dataset=dataset,  # training dataset
)

trainer.train()
```

### Error
Produces error `TypeError: unhashable type: 'slice'`. Also `/` produce floats on python3, which I guess is not intended here.
Solution in same spirit could be conditional use of
```
reduced_features = {
 ft: features[ft][:self.args.train_batch_size // self.args.n_replicas]
 for ft in features
}
```

Already mentioned here https://github.com/huggingface/transformers/pull/6038#issuecomment-664706046

### Logic issue
I don't understand what the `n_replicas` has to do with the gradient accumulation here? Shouldn't the denominator rather be `gradient_accumulation_steps`? And shouldn't it actually use the different slices of the features, and not always the "first" slice? 
Might be totally misunderstanding this.

Also this line doesn't seem to have any purpose: https://github.com/huggingface/transformers/blob/master/src/transformers/trainer_tf.py#L607

Happy to provide PR if someone can give me a hint on the logic issue.

### PR DESCRIPTION
**Pull request by jplu**
This PR fixes #6479 and fix also the NER example that was not working anymore since few weeks.

### CODE REVIEW

File: src/transformers/trainer_tf.py [hunk 1]
```diff
@@ -620,13 +620,23 @@ def apply_gradients(self, features, labels):
             self.optimizer.apply_gradients(list(zip(gradients, self.model.trainable_variables)))
         else:
             for _ in tf.range(self.args.gradient_accumulation_steps):
-                reduced_features = features[: self.args.train_batch_size / self.args.n_replicas]
-                reduced_labels = labels[: self.args.train_batch_size / self.args.n_replicas]
+                reduced_features = {
+                    ft: features[ft][: self.args.train_batch_size // self.args.n_replicas] for ft in features
```
Comment (sgugger) [inline src/transformers/trainer_tf.py:624]:
```suggestion
                    k: ft[: self.args.train_batch_size // self.args.n_replicas] for k, ft in features.items()
```

File: src/transformers/trainer_tf.py [hunk 2]
```diff
@@ -620,13 +620,23 @@ def apply_gradients(self, features, labels):
...
+                }
+                reduced_labels = labels[: self.args.train_batch_size // self.args.n_replicas]
 
                 self.training_step(reduced_features, reduced_labels)
 
-                features = tf.concat(
-                    [features[self.args.train_batch_size / self.args.n_replicas :], reduced_features], axis=0
+                features = {
+                    ft: tf.concat(
```
Comment (sgugger) [inline src/transformers/trainer_tf.py:631]: Same here, I'd loop on `features.items()`

**Review by sgugger — APPROVED** [review]
Looks great, thanks for fixing! Just have two nits.

File: src/transformers/trainer_tf.py (hunk 2 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:631]: Good to me! Doing the update.

Comment (jplu) [reply src/transformers/trainer_tf.py]: Does it looks like you proposed now?

Comment (sgugger) [reply src/transformers/trainer_tf.py]: Yup, thanks (I think this is slightly faster this way.)

Comment (jplu) [reply src/transformers/trainer_tf.py]: Perfect, looks good to merge for me then :)

**Review by LysandreJik — APPROVED** [review]
LGTM

**Pull request merged**